lib/install/management/finder.go

```diff
@@ -311,62 +327,90 @@ func (d *Dispatcher) listResourcePools(path string) ([]*object.ResourcePool, err
+					d.op.Errorf("Failed to query vapp %q: %s", vappPath, err)
+				}
+			}
+			for _, vapp := range vapps {
+				children, err := d.findVCHs(vapp.ResourcePool, multiPool)
+				if err != nil {
+					d.op.Warnf("Failed to get VCH from vApp resource pool %q: %s", pool.InventoryPath, err)
+					continue
+				}
+				vchs = append(vchs, children...)
+			}
+		}
 	}
-
 	return vchs
 }
 
-// getVCHs will check vm with same name under this resource pool, to see if that's VCH vm, and it will also check children vApp, to see if that's a VCH.
-// eventually return all fond VCH VMs
-func (d *Dispatcher) getChildVCHs(pool *object.ResourcePool, searchVapp bool) ([]*vm.VirtualMachine, error) {
+// findVCHs finds any VCH in the specified pool.  If the compute resource has multiple pools, then look
+// for any VMs with the same name as the resource pool.  If the compute resource only had a single pool then
+// evaluate each VM in the pool for VCH metadata
+func (d *Dispatcher) findVCHs(pool *object.ResourcePool, multiPool bool) ([]*vm.VirtualMachine, error) {
 	defer trace.End(trace.Begin(pool.InventoryPath, d.op))
 
 	// check if pool itself contains VCH vm.
+	var vms []*vm.VirtualMachine
 	var vchs []*vm.VirtualMachine
-	poolName := pool.Name()
+	var err error
 	computeResource := compute.NewResourcePool(d.op, d.session, pool.Reference())
-	vmm, err := computeResource.GetChildVM(d.op, d.session, poolName)
-	if err != nil {
-		return nil, errors.Errorf("Failed to query children VM in resource pool %q: %s", pool.InventoryPath, err)
-	}
-	if vmm != nil {
-		vmm.InventoryPath = path.Join(pool.InventoryPath, poolName)
-		// #nosec: Errors unhandled.
-		if ok, _ := d.isVCH(vmm); ok {
-			d.op.Debugf("%q is VCH", vmm.InventoryPath)
-			vchs = append(vchs, vmm)
-		}
-	}
 
-	if !searchVapp {
-		return vchs, nil
-	}
+	// The compute resource had multiple pools, so the assumption is that any VCH that exists will be the same
+	// name as it's resource pool
+	if multiPool {
+		vm, err := computeResource.GetChildVM(d.op, d.session, pool.Name())
+		if err != nil {
+			return nil, errors.Errorf("Failed to query children VM in resource pool %q: %s", pool.InventoryPath, err)
+		}
+		if vm != nil {
+			vms = append(vms, vm)
+		}
 
-	vappPath := path.Join(pool.InventoryPath, "*")
-	vapps, err := d.session.Finder.VirtualAppList(d.op, vappPath)
-	if err != nil {
-		if _, ok := err.(*find.NotFoundError); ok {
-			return vchs, nil
+	} else {
+		// We only had one pool, so we'll look at all the VMs in that pool
+		vms, err = computeResource.GetChildrenVMs(d.op, d.session)
+		if err != nil {
+			return nil, errors.Errorf("Failed to query children VM in resource pool %q: %s", pool.InventoryPath, err)
 		}
-		d.op.Debugf("Failed to query vapp %q: %s", vappPath, err)
 	}
-	for _, vapp := range vapps {
-		childVCHs, err := d.getChildVCHs(vapp.ResourcePool, false)
-		if err != nil {
-			return nil, err
+	// iterate over VMs and determine if we've got any VCHs
+	for _, v := range vms {
```

Comment:
Can we try a VM of the same name as the pool/folder first as in 99.9...% of cases that'll be the VCH. I don't really want to iterate over literal thousands of cVMs (real world example with selenium grid) to look at all their data.

Reply:
That will _rarely_ if ever be the case in a DRS Disabled environment -- the call to get all children should only occur in a DRS disabled environment.  So the probability that we will only have 1 VM in the slice is high.

We can make that change, but I don't think it will help.